File: bumperbot_cpp_examples/src/simple_lifecycle_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <rclcpp_lifecycle/lifecycle_node.hpp>
#include <std_msgs/msg/string.hpp>
#include <thread>

using namespace std::chrono_literals;


class SimpleLifecycleNode: public rclcpp_lifecycle::LifecycleNode
{    
public:
    SimpleLifecycleNode(const std::string & node_name, bool intra_process_comms = false)
    : LifecycleNode(node_name, rclcpp::NodeOptions().use_intra_process_comms(intra_process_comms))
    {

    }

    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn on_configure(const rclcpp_lifecycle::State & state)
    {
        RCLCPP_INFO(get_logger(), "Lifecycle Node on_configure called");
        sub_ = create_subscription<std_msgs::msg::String>(
            "chatter", 10, std::bind(&SimpleLifecycleNode::msgCallback, this, std::placeholders::_1)
        );

        return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
    }

    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn on_shutdown(const rclcpp_lifecycle::State &state)
    {
        RCLCPP_INFO(get_logger(), "Lifecycle Node on_shutdown called");
        sub_.reset();
        return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
    }

    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn on_cleanup(const rclcpp_lifecycle::State &state)
    {
        RCLCPP_INFO(get_logger(), "Lifecycle Node on_cleanup called");
        sub_.reset();
        return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
    }

    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn on_activate(const rclcpp_lifecycle::State &state)
    {
        LifecycleNode::on_activate(state);
        RCLCPP_INFO(get_logger(), "Lifecycle Node on_activate called");
        std::this_thread::sleep_for(2s);
        return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;

    }

    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn on_deactivate(const rclcpp_lifecycle::State &state)
    {
        LifecycleNode::on_deactivate(state);
        RCLCPP_INFO(get_logger(), "Lifecycle Node on_deactivate called");
        return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
    }

private:
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr sub_;
    

    void msgCallback(const std_msgs::msg::String &msg)
    {
        auto state = get_current_state();
        if(state.label() == "active")
        {
            RCLCPP_INFO_STREAM(get_logger(), "Lifecycle node heard: "<< msg.data);
        }
    }
};


int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::executors::SingleThreadedExecutor ste;
    std::shared_ptr<SimpleLifecycleNode> node = std::make_shared<SimpleLifecycleNode>("simple_lifecycle_node");
    ste.add_node(node->get_node_base_interface());
    ste.spin();
    rclcpp::shutdown();

    return 0;
}

File: bumperbot_python_examples/bumperbot_python_examples/simple_lifecycle_node.py
import rclpy
import rclpy.executors
from rclpy.lifecycle import Node, TransitionCallbackReturn, State
from std_msgs.msg import String
import time

class SimpleLifecycleNode(Node):
    def __init__(self, node_name, **kwargs):
        super().__init__(node_name, **kwargs)

    def on_configure(self, state: State) -> TransitionCallbackReturn:
        self.sub_ = self.create_subscription(String, "chatter", self.msgCallback, 10)
        self.get_logger().info("Lifecycle node on_configure called")
        return TransitionCallbackReturn.SUCCESS
    
    def on_shutdown(self, state: State) -> TransitionCallbackReturn:
        self.destroy_subscription(self.sub_)
        self.get_logger().info("Lifecycle node on_shutdown called")
        return TransitionCallbackReturn.SUCCESS
    
    def on_cleanup(self, state: State) -> TransitionCallbackReturn:
        self.destroy_subscription(self.sub_)
        self.get_logger().info("Lifecycle node on_cleanup called")
        return TransitionCallbackReturn.SUCCESS
    
    def on_activate(self, state: State) -> TransitionCallbackReturn:
        self.get_logger().info("Lifecycle node on_activate called")
        time.sleep(2)
        return super().on_activate(state)
    
    def on_deactivate(self, state: State) -> TransitionCallbackReturn:
        self.get_logger().info("Lifecycle node on_deactivate called")
        return super().on_deactivate(state)
    
    def msgCallback(self, msg: String):
        current_state = self._state_machine.current_state
        if current_state[1] == "active":
            self.get_logger().info(f"I heard: {msg.data}")




def main():
    rclpy.init()
    executor = rclpy.executors.SingleThreadedExecutor()
    simple_lifecycle_node = SimpleLifecycleNode("simple_lifecycle_node")
    executor.add_node(simple_lifecycle_node)
    try:
        executor.spin()
    except(KeyboardInterrupt, rclpy.executors.ExternalShutdownException):
        simple_lifecycle_node.destroy_node()

    rclpy.shutdown()

if __name__ == "__main__":
    main()

